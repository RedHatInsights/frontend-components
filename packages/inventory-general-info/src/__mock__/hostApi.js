import { hosts } from '../api';
import MockAdapter from 'axios-mock-adapter';

export const mock = new MockAdapter(hosts.axios);
