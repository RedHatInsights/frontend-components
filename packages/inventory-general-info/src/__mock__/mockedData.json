{
    "total": 1,
    "count": 1,
    "page": 1,
    "per_page": 50,
    "results": [
        {
            "id": "6de30d78-f052-41d4-b123-125017443a98",
            "system_profile": {
                "arch": "x86_64",
                "cpu_flags": [ "fpu" ],
                "yum_repos": [
                    {
                        "name": "Extra Packages for Enterprise Linux 7 - Testing - $basearch",
                        "enabled": false,
                        "gpgcheck": true
                    }
                ],
                "os_release": "7.4",
                "bios_vendor": "SeaBIOS",
                "bios_version": "1.10.2-2.fc27",
                "kernel_modules": [ "veth" ],
                "last_boot_time": "2018-03-14T17:06:10",
                "number_of_cpus": 1,
                "cores_per_socket": 1,
                "enabled_services": [ "auditd" ],
                "bios_release_date": "04\/01\/2014",
                "number_of_sockets": 1,
                "os_kernel_version": "3.10.0",
                "running_processes": [ "qmgr" ],
                "satellite_managed": false,
                "installed_packages": [ "0:GeoIP-1.5.0-11.el7" ],
                "installed_services": [ "arp-ethers" ],
                "network_interfaces": [
                    {
                        "mtu": 65536,
                        "name": "lo",
                        "type": "loopback",
                        "state": "UNKNOWN",
                        "mac_address": "00:00:00:00:00:00",
                        "ipv4_addresses": [
                            "127.0.0.1"
                        ],
                        "ipv6_addresses": [
                            "::1"
                        ]
                    }
                ],
                "infrastructure_type": "virtual",
                "system_memory_bytes": 512073728,
                "infrastructure_vendor": "qemu"
            }
        }
    ]
}
