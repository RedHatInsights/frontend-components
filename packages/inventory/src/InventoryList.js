import React from 'react';
import { connect } from 'react-redux';
import InventoryEntityTable from './EntityTable';
import { loadEntities, showEntities, clearFilters } from './redux/actions';
import { Grid, GridItem } from '@patternfly/react-core';
import PropTypes from 'prop-types';
import './InventoryList.scss';
import { InventoryContext } from './Inventory';
import { CancelToken } from 'axios';

class ContextInventoryList extends React.Component {
    constructor(props) {
        super(props);
    }

    loadEntities = (options = {}, reload = true) => {
        const { page, perPage, onRefresh, items, hasItems, sortBy } = this.props;
        options = {
            page: options.page || page,
            // eslint-disable-next-line camelcase
            per_page: options.per_page || perPage,
            orderBy: sortBy && sortBy.key,
            orderDirection: sortBy && sortBy.direction.toUpperCase(),
            ...options
        };
        reload && onRefresh(options);
        if (this.controller) {
            this.controller.cancel('Get host items canceled by user.');
        }

        this.controller = CancelToken.source();
        this.props.loadEntities && this.props.loadEntities(
            items,
            {
                ...options,
                controller: this.controller,
                prefix: this.props.pathPrefix,
                base: this.props.apiBase,
                hasItems
            }
        );
    }

    componentDidMount() {
        const { setRefresh, setUpdate, onClearFilters } = this.props;
        setRefresh && setRefresh(this.loadEntities);
        setUpdate && setUpdate((options) => this.loadEntities(options, false));
        this.loadEntities();
        onClearFilters();
    }

    componentDidUpdate(prevProps) {
        const { items, hasItems, sortBy } = this.props;
        if (hasItems && JSON.stringify(items) !== JSON.stringify(prevProps.items)) {
            this.loadEntities({}, false);
        } else if (!hasItems && JSON.stringify(prevProps.sortBy) !== JSON.stringify(sortBy)) {
            this.loadEntities({}, false);
        }
    }

    render() {
        const { showHealth, ...props } = this.props;
        return (
            <React.Fragment>
                <Grid guttter="sm" className="ins-inventory-list">
                    <GridItem span={ 12 }>
                        <InventoryEntityTable { ...props } showHealth={ showHealth } />
                    </GridItem>
                </Grid>
            </React.Fragment>
        );
    }
}

const propTypes = {
    filterEntities: PropTypes.func,
    loadEntities: PropTypes.func,
    pathPrefix: PropTypes.number,
    apiBase: PropTypes.string,
    showHealth: PropTypes.bool,
    page: PropTypes.number,
    perPage: PropTypes.number,
    onRefresh: PropTypes.func,
    sortBy: PropTypes.shape({
        key: PropTypes.string,
        direction: PropTypes.string
    }),
    items: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.string),
        PropTypes.shape({
            id: PropTypes.string.isRequired
        }),
        PropTypes.shape({
            account: PropTypes.any,
            isOpen: PropTypes.bool,
            title: PropTypes.node
        })
    ]),
    entities: PropTypes.arrayOf(PropTypes.any)
};

ContextInventoryList.propTypes = {
    ...propTypes,
    setRefresh: PropTypes.func
};

ContextInventoryList.defaultProps = {
    perPage: 50,
    page: 1,
    onRefresh: () => undefined
};

const InventoryList = ({ ...props }) => (
    <InventoryContext.Consumer>
        { ({ setRefresh, setUpdate }) => (
            <ContextInventoryList { ...props } setRefresh={ setRefresh } setUpdate={ setUpdate } />
        ) }
    </InventoryContext.Consumer>
);

InventoryList.propTypes = propTypes;

function mapDispatchToProps(dispatch) {
    return {
        onClearFilters: () => dispatch(clearFilters()),
        loadEntities: (items = [], config) => {
            if (!Array.isArray(items)) {
                console.error('Wrong shape of items, array with strings or objects with ID property required!');
            }

            const limitedItems = items.slice((config.page - 1) * config.per_page, config.page * config.per_page);

            if (limitedItems.length > 0) {
                config.itemsPage = config.page;
                config.page = 1;
            }

            const itemIds = limitedItems.reduce((acc, curr) => (
                [
                    ...acc,
                    curr && typeof curr === 'string' ? curr : curr.id
                ]
            ), []).filter(Boolean);
            dispatch(loadEntities(itemIds, config));
            dispatch(showEntities(limitedItems.map(oneItem => (
                { ...typeof oneItem === 'string' ? { id: oneItem } : oneItem }
            ))));
        }
    };
}

export default connect(
    ({ entities: { page, perPage, sortBy }}, { perPage: currPerPage }) => ({ page, perPage: currPerPage || perPage, sortBy }),
    mapDispatchToProps
)(InventoryList);
