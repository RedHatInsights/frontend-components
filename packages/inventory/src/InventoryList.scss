@import "~@patternfly/patternfly/utilities/Flex/flex.css";
@import "~@patternfly/patternfly/utilities/Display/display.css";
@import '~@redhat-cloud-services/frontend-components-utilities/files/Utilities/_all';
@import '~@redhat-cloud-services/frontend-components/components/TagCount.css';
@import '~@redhat-cloud-services/frontend-components/components/TagModal.css';

.ins-c-inventory__table--toolbar {
    .pf-c-data-toolbar__content {
        .pf-c-data-toolbar__item:last-child { @include rem('min-width', 200px); }
    }
    .pf-c-pagination__nav-page-select input { @include rem('min-width', 75px); }
}

.ins-c-table__empty {
    @include rem('margin', 75px);
    border-bottom: none;

    .pf-c-content {
        text-align: center;
    }
}

.ins-c-entity-table.pf-c-table {

    .ins-composed-col div:nth-child(2) {
        color: var(--pf-global--link--Color);
        word-wrap: break-word;
        word-break:break-all;
        cursor: pointer;
        &:hover {
            color: var(--pf-global--link--Color--hover);
        }
    }

    .ins-health-status div {
        display: inline-block;
        padding-right: 20px;

        svg {
            position: relative;
            top: 2px;
            left: -2px;
        }
    }

    .ins-composed-col {
        :first-child {
            color: var(--pf-c-button--m-disabled--Color);
            font-size: var(--pf-global--FontSize--sm);
        }

        * {
            color: var(--pf-global--info-color--100);
            font-size: var(	--pf-global--FontSize--lg);
        }

        .ins-composed-col {
            cursor: pointer;
        }
    }

    .ins-health-status div {
        span:nth-child(2) {
            color: var(--pf-global--info-color--100);
            cursor: pointer;
        }
    }

    .pf-c-table__action .pf-c-dropdown {
        position: relative;
        left: 10px;
        top: 3px;
    }
}

.ins-simple-table {
    .ins-health-status div {
        display: inline-block;
        padding-right: 20px;

        svg {
          position: relative;
            top: 2px;
            left: -2px;
        }
    }
}
