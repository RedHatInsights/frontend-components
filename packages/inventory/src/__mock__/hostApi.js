import { hosts, tags } from '../api';
import MockAdapter from 'axios-mock-adapter';

export const mock = new MockAdapter(hosts.axios);
export const mockTags = new MockAdapter(tags.axios);
