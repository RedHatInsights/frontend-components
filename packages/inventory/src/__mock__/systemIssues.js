import { systemIssuesInstance } from '../api';
import MockAdapter from 'axios-mock-adapter';

export const mock = new MockAdapter(systemIssuesInstance);
