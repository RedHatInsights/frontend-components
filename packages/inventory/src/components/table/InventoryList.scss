@import "~@patternfly/patternfly/sass-utilities/mixins.scss";
@import "~@patternfly/patternfly/sass-utilities/colors.scss";
@import "~@patternfly/patternfly/sass-utilities/scss-variables.scss";
@import "~@patternfly/patternfly/utilities/Flex/flex.scss";
@import "~@patternfly/patternfly/utilities/Display/display.scss";
@import '~@redhat-cloud-services/frontend-components-utilities/files/Utilities/_all';
@import '~@redhat-cloud-services/frontend-components/components/TagCount.css';
@import '~@redhat-cloud-services/frontend-components/components/TagModal.css';
@import '~@redhat-cloud-services/frontend-components/components/CullingInfo.css';

.ins-c-inventory__table--toolbar {
    &.ins-m-footer {
        margin: 0;
        padding: 0;
    }

    &:not(.ins-c-inventory__table--toolbar-has-items) {
        .pf-c-toolbar__content {
            .pf-c-toolbar__item:last-child { @include rem('min-width', 200px); }
        }
        .pf-c-pagination__nav-page-select input { @include rem('min-width', 75px); }
    }

    .ins-c-chip-filters .pf-c-chip-group:not(:last-of-type) { margin-right: var(--pf-global--spacer--xs); }
}

.ins-c-table__empty {
    @include rem('margin', 75px);
    border-bottom: none;

    .pf-c-content {
        text-align: center;
    }
}

.ins-c-inventory__tags-filter ul {
    label.pf-c-check__label {
        display: flex;
        width: 100%;

        .ins-c-inventory__tags-filter--tag-name {
            margin-right: var(--pf-global--spacer--xs);
        }
        .pf-c-badge { margin-left: auto; }
    }

    .ins-c-inventory__tags-tail {
        display: flex;

        .ins-c-inventory__tags-no-tags { margin: auto; }

        .pf-c-spinner { margin-left: auto; }
    }
}

.ins-c-entity-table.pf-c-table {

    .pf-c-table__check input { cursor: pointer; }
    .ins-composed-col div:nth-child(2) {
        color: var(--pf-global--link--Color);
        word-wrap: break-word;
        word-break:break-all;
        cursor: pointer;
        &:hover {
            color: var(--pf-global--link--Color--hover);
        }
    }

    .ins-health-status div {
        display: inline-block;
        padding-right: 20px;

        svg {
            position: relative;
            top: 2px;
            left: -2px;
        }
    }

    .ins-composed-col {
        > :first-child {
            color: var(--pf-c-button--m-disabled--Color);
            font-size: var(--pf-global--FontSize--sm);
        }

        * {
            font-size: var(	--pf-global--FontSize--lg);
        }

        .ins-composed-col {
            cursor: pointer;
        }
    }

    .ins-c-inventory__list-tags {
        display: inline-block;
    }

    .ins-health-status div {
        span:nth-child(2) {
            color: var(--pf-global--info-color--100);
            cursor: pointer;
        }
    }

    .pf-c-table__action .pf-c-dropdown {
        position: relative;
        left: 10px;
        top: 3px;
    }
}

.ins-simple-table {
    .ins-health-status div {
        display: inline-block;
        padding-right: 20px;

        svg {
          position: relative;
            top: 2px;
            left: -2px;
        }
    }
}
