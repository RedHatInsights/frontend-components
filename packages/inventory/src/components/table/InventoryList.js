import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import InventoryEntityTable from './EntityTable';
import { Grid, GridItem } from '@patternfly/react-core/dist/esm/layouts/Grid';
import PropTypes from 'prop-types';
import './InventoryList.scss';
import { loadSystems } from '../../shared';
import isEqual from 'lodash/isEqual';

/**
 * Component that works as a side channel for consumers to notify inventory of new data changes.
 */
class ContextInventoryList extends React.Component {
    /**
     * If conumer wants to change data they can call this function via component ref.
     * @param {*} options new options to be applied, like pagination, filters, etc.
     */
    onRefreshData = (options = {}) => {
        const { page, perPage, items, hasItems, sortBy, activeFilters, showTags } = this.props;
        this.props.loadEntities && this.props.loadEntities({
            page,
            perPage,
            items,
            hasItems,
            sortBy,
            activeFilters,
            ...options
        }, showTags);
    }

    /**
     * Function to calculate for new changes, this function limits re-renders by checking if previous items are
     * same as new items.
     * If items are not passed, it only checks for props sortBy.
     * @param {*} prevProps previous props - items, hasItems, sortBy.
     */
    componentDidUpdate(prevProps) {
        const { items, hasItems, sortBy } = this.props;
        if (
            hasItems &&
            !isEqual(
                items.map(({ children, isOpen, ...item }) => item),
                prevProps.items.map(({ children, isOpen, ...item }) => item)
            )
        ) {
            this.onRefreshData({});
        } else if (!hasItems && !isEqual(prevProps.sortBy, sortBy)) {
            this.onRefreshData({});
        }
    }

    componentDidMount() {
        this.onRefreshData({});
    }

    render() {
        const { showHealth, ...props } = this.props;
        return (
            <Grid guttter="sm" className="ins-inventory-list">
                <GridItem span={ 12 }>
                    <InventoryEntityTable { ...props } />
                </GridItem>
            </Grid>
        );
    }
}

/**
 * Component that consumes active filters and passes them down to component.
 */
const InventoryList = React.forwardRef((props, ref) => {
    const dispatch = useDispatch();
    const activeFilters = useSelector(({ entities: { activeFilters } }) => activeFilters);
    return (
        <ContextInventoryList
            { ...props }
            ref={ref}
            activeFilters={ activeFilters }
            loadEntities={ (config, showTags) => dispatch(loadSystems(config, showTags)) }
        />
    );
});

ContextInventoryList.propTypes = {
    ...InventoryList.propTypes,
    setRefresh: PropTypes.func
};
ContextInventoryList.defaultProps = {
    perPage: 50,
    page: 1
};
InventoryList.propTypes = {
    showTags: PropTypes.bool,
    filterEntities: PropTypes.func,
    loadEntities: PropTypes.func,
    showHealth: PropTypes.bool,
    page: PropTypes.number,
    perPage: PropTypes.number,
    sortBy: PropTypes.shape({
        key: PropTypes.string,
        direction: PropTypes.string
    }),
    items: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.string),
        PropTypes.shape({
            id: PropTypes.string.isRequired
        }),
        PropTypes.shape({
            account: PropTypes.any,
            isOpen: PropTypes.bool,
            title: PropTypes.node
        })
    ]),
    entities: PropTypes.arrayOf(PropTypes.any)
};

export default InventoryList;
