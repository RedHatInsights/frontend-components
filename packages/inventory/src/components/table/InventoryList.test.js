/* eslint-disable react/display-name */
import React from 'react';
import InventoryList from './InventoryList';
import { render, mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import promiseMiddleware from 'redux-promise-middleware';
import toJson from 'enzyme-to-json';
import { mock } from '../../__mock__/hostApi';

describe('InventoryList', () => {
    let initialState;
    let mockStore;

    beforeEach(() => {
        mockStore = configureStore([ promiseMiddleware() ]);
        initialState = {
            entities: {
                activeFilters: [{}],
                loaded: true,
                rows: [],
                columns: [{ key: 'one', title: 'One' }],
                page: 1,
                perPage: 50,
                total: 500
            }
        };
    });

    it('should render correctly', () => {
        const store = mockStore(initialState);
        const wrapper = render(<Provider store={ store }>
            <InventoryList/>
        </Provider>);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    describe('API', () => {
        it('should fire refresh after changing sort', () => {
            mock.onGet('/api/inventory/v1/hosts').reply(200, {});
            const sortBy = {
                index: 1,
                key: 'one',
                direction: 'asc'
            };
            const store = mockStore(initialState);
            const Cmp = (props) => <Provider store={ store }>
                <InventoryList {...props} />
            </Provider>;
            const wrapper = mount(<Cmp sortBy={sortBy} />);
            wrapper.setProps({
                sortBy: {
                    ...sortBy,
                    direction: 'desc'
                }
            });
            wrapper.update();
            const actions = store.getActions();
            expect(actions[0].type).toBe('LOAD_ENTITIES_PENDING');
        });

        it('should not fire refresh after changing props', () => {
            const sortBy = {
                index: 1,
                key: 'one',
                direction: 'asc'
            };
            const store = mockStore(initialState);
            const Cmp = (props) => <Provider store={ store }>
                <InventoryList {...props} />
            </Provider>;
            const wrapper = mount(<Cmp sortBy={sortBy} />);
            wrapper.setProps({
                showTags: true
            });
            wrapper.update();
            const actions = store.getActions();
            expect(actions.length).toBe(0);
        });

        it('should fire refresh after changing items', () => {
            mock.onGet(/\/api\/inventory\/v1\/hosts.*/).reply(200, {});
            const store = mockStore(initialState);
            const Cmp = (props) => <Provider store={ store }>
                <InventoryList {...props} />
            </Provider>;
            const wrapper = mount(<Cmp items={[{ children: () => <div>test</div>, isOpen: false, id: 'fff' }]} hasItems />);
            wrapper.setProps({
                items: [{ children: () => <div>test</div>, isOpen: false, id: 'something' }]
            });
            wrapper.update();
            const actions = store.getActions();
            expect(actions[0].type).toBe('LOAD_ENTITIES_PENDING');
        });
    });
});
