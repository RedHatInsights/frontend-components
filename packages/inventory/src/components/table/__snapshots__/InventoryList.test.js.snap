// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InventoryList should render correctly 1`] = `
<div
  class="pf-l-grid ins-inventory-list"
  guttter="sm"
>
  <div
    class="pf-l-grid__item pf-m-12-col"
  >
    <table
      aria-label="Host inventory"
      class="pf-c-table pf-m-grid-md ins-c-entity-table"
      data-ouia-component-id="0"
      data-ouia-component-type="PF4/Table"
      data-ouia-safe="true"
      role="grid"
    >
      <thead
        class=""
      >
        <tr>
          <th
            class=""
            data-key="0"
            data-label="One"
            scope="col"
          >
            One
          </th>
        </tr>
      </thead>
      <tbody
        class=""
      >
        <tr
          class=""
          data-ouia-component-id="1"
          data-ouia-component-type="PF4/TableRow"
          data-ouia-safe="true"
        >
          <td
            class=""
            colspan="1"
            data-key="0"
            data-label="One"
          >
            <div
              class="ins-c-table__empty"
            >
               
              <div
                class="pf-l-bullseye"
              >
                <div
                  class="pf-c-empty-state"
                >
                  <div
                    class="pf-c-empty-state__content"
                  >
                    <h5
                      class="pf-c-title pf-m-lg"
                    >
                      No matching systems found
                    </h5>
                    <div
                      class="pf-c-empty-state__body"
                    >
                      This filter criteria matches no systems. 
                      <br />
                       Try changing your filter settings.
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
